relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+            .finish()
+    }
+}
+
+/// A collection of weighted [app features](AppFeature).
+///
+/// Used to attribute a single COGS measurement to multiple features.
+#[derive(Clone, Copy)]
+pub struct AppFeatures(EnumMap<16, AppFeature, NonZeroUsize>);
+
+impl AppFeatures {
+    /// Attributes all measurements to a single [`AppFeature`].
+    pub fn new(feature: AppFeature) -> Self {
+        Self::builder().weight(feature, 1).build()
+    }
+
+    /// Attributes the measurement to nothing.
+    pub fn none() -> Self {
+        Self::builder().build()
+    }
+
+    /// Returns an [`AppFeatures`] builder.
+    pub fn builder() -> AppFeaturesBuilder {
+        AppFeaturesBuilder(Self(Default::default()))
+    }
+
+    /// Merges two instances of [`AppFeatures`] and sums the contained weights.
+    pub fn merge(mut self, other: Self) -> Self {
+        for (feature, weight) in other.0.into_iter() {
+            if let Some(w) = self.0.get_mut(feature) {
+                *w = w.saturating_add(weight.get());
+            } else {
+                self.0.insert(feature, weight);
+            }
+        }
+
+        self
+    }
+
+    /// Returns an iterator yielding an app feature and it's associated weight
+    /// normalized to the total stored weights in the range between `0.0` and `1.0`.
+    ///
+    /// Used to divide a measurement by the stored weights.
+    pub fn weights(&self) -> impl Iterator<Item = (AppFeature, f32)> {
+        let total_weight: usize = self.0.into_iter().map(|(_, weight)| weight.get()).sum();
+
+        self.0.into_iter().filter_map(move |(feature, weight)| {
+            if total_weight == 0 {
+                return None;
+            }
+
+            let ratio = (weight.get() as f32 / total_weight as f32).clamp(0.0, 1.0);
+            Some((feature, ratio))
+        })
+    }
+}
+
+impl fmt::Debug for AppFeatures {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "AppFeatures(")?;
+
+        let mut first = true;
+        for (feature, weight) in self.weights() {
+            if !first {
+                first = false;
+                write!(f, ", ")?;
+            }
+            write!(f, "{feature:?}={weight:.2}")?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl From<AppFeature> for AppFeatures {
+    fn from(value: AppFeature) -> Self {
+        Self::new(value)
+    }
+}
+
+/// A builder for [`AppFeatures`] which can be used to configure different weights per [`AppFeature`].
+pub struct AppFeaturesBuilder(AppFeatures);
+
+impl AppFeaturesBuilder {
+    /// Increases the `weight` of an [`AppFeature`].
+    pub fn add_weight(mut self, feature: AppFeature, weight: usize) -> Self {
+        let Some(weight) = NonZeroUsize::new(weight) else {
+            return self;
+        };
+
+        if let Some(previous) = self.0 .0.get_mut(feature) {
+            *previous = previous.saturating_add(weight.get());
+        } else {
+            self.0 .0.insert(feature, weight);
```

Comment:
```suggestion
            self.0.0.insert(feature, weight);
```

Reply:
For some reason cargo format formats it like this, it doesn't even accept the suggestion and corrects it.